FakeTensor speedup: c++ extract_tensor_metadata()

aorenste opened the pull request:
An implementation of `extract_tensor_metadata()` in C++ for computing the cache key.

Instead of building a `TensorMetadata` the C++ code just builds a string which can be used for caching but isn't reversible.

19.0% FakeTensor perf win on the microbenchmark (39.4% cumulative)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125312
* #124226
* #124225

aorenste committed:
WIP: c++ fake_tensor cache_key

[ghstack-poisoned]

aorenste committed:
Update on "WIP: c++ fake_tensor cache_key"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: c++ fake_tensor cache_key"




[ghstack-poisoned]

aorenste committed:
Update on "WIP: c++ fake_tensor cache_key"




[ghstack-poisoned]

ezyang commented:
deferring to @masnesral

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1190,7 +1198,7 @@ def _prep_args_for_hash(self, output: List[Any], args: Any):
                         output.append(_UNHASHABLE)
                     else:
-                        output.append(extract_tensor_metadata(arg))
+                        output.append(torch._C._FakeTensor_extract_tensor_metadata(arg))
```
nit: perhaps the name should capture the nuance that this metadata is only useful for hashing purposes? I dunno; I don't feel strongly ¯\_(ツ)_/¯

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -759,10 +759,18 @@ class TensorMetadata:
     Extract the TensorMetadata of a tensor.
     """
+
+    # WARNING: Any changes to this (probably) need to be reflected in
```
I wonder if there's some kind of unit test we could write to alert us if someone adds a field and does not keep them in sync? I guess nothing clever comes to mind.

Actually, all the existing unit tests that validate different hashes for different FakeTensor properties will hit the new codepath, correct? So ideally someone should add a new test for any new field and it would fail if there's not a corresponding change in the cpp?

aorenste replied:
Do we have some kind of "extra debug" flag we can check on the python side? We store the extract_tensor_metadata() in the cache entry - so we could always verify that for a given cache value the metadata  hasn't changed - but I'd only want to do that in some kind of "paranoid" mode since it would defeat a bunch of the perf savings.

aorenste replied:
Oh - I suppose I could attach it to torch._dynamo.config.fake_tensor_cache_crosscheck_enabled which already does a bunch of validation...

masnesral replied:
Oh yeah, that might be a good idea.

masnesral commented on torch/csrc/tensor/fake_tensor.cpp:
```diff
@@ -0,0 +1,273 @@
+const std::array<int, 4> DIM_ORDER_4{1, 3, 2, 0};
+const std::array<int, 5> DIM_ORDER_5{1, 4, 3, 2, 0};
+
+bool are_strides_like_channels_last(const at::Tensor& tensor) {
```
If you're looking for a better review on this part, lemme know. I never looked deeply into channels last format. So I have no idea. Is this the sort of thing that deserves some unit tests? I don't know what precedent we have for testing c++ bindings like this.

masnesral replied:
I guess the same for the functionality below that gets into the "internals" like is_sparse

masnesral reviewed (approved):
Well.... I'm not at all familiar with any of the cpp code in torch including the cpp bindings so I don't know if I'm the best to give useful feedback here. FWIW, the impl looks pretty good to me. My only hesitation is a general unease about keeping this in sync with other parts of the code base. New tensor metadata fields are one case that you've already pointed out. But what if someone adds a new sparse format, for example? Incorrect cache hits from something like that would be a bummer.

eellison commented on torch/csrc/tensor/fake_tensor.cpp:
```diff
@@ -0,0 +1,273 @@
+const std::array<int, 4> DIM_ORDER_4{1, 3, 2, 0};
+const std::array<int, 5> DIM_ORDER_5{1, 4, 3, 2, 0};
+
+bool are_strides_like_channels_last(const at::Tensor& tensor) {
+  const auto& strides = tensor.sym_strides();
```
there should an existing method for this: `suggest_memory_format` on TensorBase

eellison commented on torch/csrc/tensor/fake_tensor.cpp:
```diff
@@ -0,0 +1,273 @@
+}
+
+bool is_sparse_any(const at::Tensor& tensor) {
+  // return is_sparse_coo(tensor) || is_sparse_compressed(tensor);
+  // THPVariable_layout
```
leftover ?

aorenste replied:
I purposely left those comments in there because looking up the dispatch -> C++ function isn't always so obvious so I left it as a hint. I'll remove them.

eellison commented on torch/csrc/tensor/fake_tensor.cpp:
```diff
@@ -0,0 +1,273 @@
+  }
+
+  if (are_strides_like_channels_last(tensor)) {
+    // THPVariable_get_ndim
```
itnentional?

eellison reviewed (commented):
didnt do full review

aorenste committed:
Update on "FakeTensor speedup: c++ extract_tensor_metadata()"

An implementation of `extract_tensor_metadata()` in C++ for computing the cache key.

Instead of building a `TensorMetadata` the C++ code just builds a string which can be used for caching but isn't reversible.

19.0% FakeTensor perf win on the microbenchmark (39.4% cumulative)




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: c++ extract_tensor_metadata()"

An implementation of `extract_tensor_metadata()` in C++ for computing the cache key.

Instead of building a `TensorMetadata` the C++ code just builds a string which can be used for caching but isn't reversible.

19.0% FakeTensor perf win on the microbenchmark (39.4% cumulative)




[ghstack-poisoned]

masnesral commented:
So thinking a little more... not sure if we have precedent for how to set the bar, but (generally speaking) any c++ bindings are a bigger commitment in terms of complexity and future support vs. pure Python. Do you think we should we benchmark against our existing suites first to see if there's a meaningful improvement on any of them before committing this?

eellison commented:
also, maybe worth implementing symint caching before moving this to C++
